### ISSUE TITLE
Grammar rules

### ISSUE THREAD

**Issue by Gipson62** [RedGear-Studio/Reg-Lang#17]
Current grammar (subject to modification):
```haskell
program = { SOI ~ (statement ~ ";")* ~ EOI }
WHITESPACE = _{ " " | "\t" | "\r" | "\n" }

statement = { variable_declaration | assignment | print_statement | if_statement | while_loop }
variable_declaration = { "let" ~ identifier ~ ":" ~ type ~ ("=" ~ expression)? }
assignment = { identifier ~ "=" ~ expression }
print_statement = { "print" ~ expression }
if_statement = { "if" ~ expression ~ "then" ~ (statement ~ ";")* ~ ("else" ~ (if_statement | statement ~ ";"))? ~ "end" }
while_loop = { "while" ~ expression ~ "do" ~ (statement ~ ";")* ~ "end" }

expression = { literal | binary_expression | identifier | unary_expression }

binary_expression = { ^"(" ~ expression ~ operator ~ expression ~ ^")"}
unary_expression = { unary_operator ~ expression }

unary_operator = { "-" }

operator = { add | sub | mul | div | mod | eq | neq | lt | lte | gt | gte }
add = _{ ^"+" }
sub = _{ ^"-" }
mul = _{ ^"*" }
div = _{ ^"/" }
mod = _{ ^"%" }
eq = _{ ^"==" }
neq = _{ ^"!=" }
lt = _{ ^"<" }
lte = _{ ^"<=" }
gt = _{ ^">" }
gte = _{ ^">=" }

literal = { number | string | boolean }
number = @{ digit+ }
string = _{ "\"" ~ string_value ~ "\"" }
string_value = { character* }
boolean = { "true" | "false" }

type = { "int" | "string" | "boolean" }

identifier = { letter ~ ( letter | digit )* }

letter = _{ 'a'..'z' | 'A'..'Z' }
digit = _{ '0'..'9' }
character = _{ letter | digit | " " | "." | "," | "!" | "?" | ";" | ":" }```

**Comment by Gipson62** [comment]
What's the syntax currently looked like based on the grammar:
# Statements
## Variable declaration:
`let IDENTIFIER: type = EXPRESSION;`
> ("`= EXPRESSION`" is optional and "`EXPRESSION`" itself could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`IDENTIFIER`" is a word who contains letters or number however long you wants but it needs to start with a letter at least)

## Assignment:
`IDENTIFIER = EXPRESSION;`
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`IDENTIFIER`" is a word who contains letters or number however long you wants but it needs to start with a letter at least)

## Print statement:
`print EXPRESSION`
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)

## If statement:
```basic
if EXPRESSION then
  STATEMENT(S)
else
  STATEMENT(S)
end;
```
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`STATEMENT(S)`" isn't mandatory and could be a variable declaration, an assignment, a print statement, another if statement or a while loop)
> (The "`else`" part isn't mandatory either, the "`end;`" in the other hand is, the compiler needs to know where the "if" ends)

## While loop: 
```basic
while EXPRESSION do
  STATEMENT(S)
end;
```
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`STATEMENT(S)`" isn't mandatory and could be a variable declaration, an assignment, a print statement, if statement or another while loop)
> (The "`end;`" is mandatory, the compiler needs to know where the "if" ends)

# Expressions:
## Binary Operation:
`(EXPRESSION OPERATOR EXPRESSION)` (e.g. `(5 + (5 * 5))`)
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`OPERATOR`" is one of these: "`+`", "`-`", "`*`", "`/`", "`%`", "`==`", "`!=`", "`<`", "`<=`", "`>`", "`>=`")
> (For now we need to use parentheses because of the parser architecture, but later they'll be removed)

## Unary Operation:
`UNARY_OPERATOR EXPRESSION` (e.g. `-(5 + 5)` | `-5`)
> ("`EXPRESSION`" could be a number, a string, an identifier, a binary expression and a unary expression)
> ("`UNARY_OPERATOR`" can (for now) only be minus, there aren't any other supported unary operator for now)

## Identifier:
`variable_name`
> A word who contains letters or number however long you wants but it needs to start with a letter at least

## Literals:
### Number:
`5`
> Only `integer`, `floats` aren't implemented as I'm sending this update.

### String:
`"The string goes here"`

### Boolean
Either `true` or `false`

**Comment by Gipson62** [comment]
Add `block` rules: `block = { (statement ~ semicolon)* }`
> `semicolon` as also been added as a silent rule to make more accurate error handling
The block rule is used in if and while statement. It's used to separate for example the if block of the else block

**Comment by Gipson62** [comment]
This was an issue for the v0.0.1 of the grammar, so it's fixed with #18
